--// Imports \\--
local Codec = require(script.Parent.Parent.Parent.Codec)

-- Wire: u16(enumTypeId) + LEB128(value).
-- typeId is an engine reflection id (NOT Enum:GetEnums() order).
local TYPE_ID_TO_ENUM: { [number]: Enum } = {
	[0] = Enum.AccessModifierType,
	[1] = Enum.AccessoryType,
	[2] = Enum.ActionType,
	[3] = Enum.ActivePayerStatus,
	[4] = Enum.ActuatorRelativeTo,
	[5] = Enum.ActuatorType,
	[6] = Enum.AdAvailabilityResult,
	[7] = Enum.AdEventType,
	[8] = Enum.AdFormat,
	[9] = Enum.AdShape,
	[10] = Enum.AdTeleportMethod,
	[11] = Enum.AdUIEventType,
	[12] = Enum.AdUIType,
	[13] = Enum.AdUnitStatus,
	[14] = Enum.AdornCullingMode,
	[15] = Enum.AdornShading,
	[16] = Enum.AgeCheckStatus,
	[17] = Enum.AlignType,
	[18] = Enum.AlphaMode,
	[19] = Enum.AnalyticsCustomFieldKeys,
	[20] = Enum.AnalyticsEconomyAction,
	[21] = Enum.AnalyticsEconomyFlowType,
	[22] = Enum.AnalyticsEconomyTransactionType,
	[23] = Enum.AnalyticsLogLevel,
	[24] = Enum.AnalyticsProgressionStatus,
	[25] = Enum.AnalyticsProgressionType,
	[26] = Enum.AnimationClipFromVideoStatus,
	[27] = Enum.AnimationNodeBlend2DInputMode,
	[28] = Enum.AnimationNodeInterruptible,
	[29] = Enum.AnimationNodePhaseSync,
	[30] = Enum.AnimationNodePlayMode,
	[31] = Enum.AnimationNodeTransitionType,
	[32] = Enum.AnimationNodeType,
	[33] = Enum.AnimationNodeWaitFor,
	[34] = Enum.AnimationPriority,
	[35] = Enum.AnimatorRetargetingMode,
	[41] = Enum.AntiAliasing,
	[42] = Enum.AppLifecycleManagerState,
	[43] = Enum.AppShellActionType,
	[44] = Enum.AppShellFeature,
	[45] = Enum.AppUpdateStatus,
	[46] = Enum.ApplyStrokeMode,
	[47] = Enum.AspectType,
	[48] = Enum.AssetCreatorType,
	[49] = Enum.AssetFetchStatus,
	[50] = Enum.AssetRepresentation,
	[51] = Enum.AssetType,
	[52] = Enum.AssetTypeVerification,
	[53] = Enum.AudioApiRollout,
	[55] = Enum.AudioChannelLayout,
	[56] = Enum.AudioFilterType,
	[57] = Enum.AudioSimulationFidelity,
	[58] = Enum.AudioSubType,
	[59] = Enum.AudioWindowSize,
	[60] = Enum.AuthorityMode,
	[61] = Enum.AutomaticSize,
	[62] = Enum.AvatarAssetType,
	[63] = Enum.AvatarChatServiceFeature,
	[64] = Enum.AvatarContextMenuOption,
	[65] = Enum.AvatarGenerationError,
	[66] = Enum.AvatarItemType,
	[67] = Enum.AvatarPromptResult,
	[68] = Enum.AvatarSettingsAccessoryLimitMethod,
	[69] = Enum.AvatarSettingsAccessoryMode,
	[70] = Enum.AvatarSettingsAnimationClipsMode,
	[71] = Enum.AvatarSettingsAnimationPacksMode,
	[72] = Enum.AvatarSettingsAppearanceMode,
	[73] = Enum.AvatarSettingsBuildMode,
	[74] = Enum.AvatarSettingsCharacterControllerMode,
	[75] = Enum.AvatarSettingsClothingMode,
	[76] = Enum.AvatarSettingsCollisionMode,
	[77] = Enum.AvatarSettingsCustomAccessoryMode,
	[78] = Enum.AvatarSettingsCustomBodyType,
	[79] = Enum.AvatarSettingsCustomClothingMode,
	[80] = Enum.AvatarSettingsHitAndTouchDetectionMode,
	[81] = Enum.AvatarSettingsJumpMode,
	[82] = Enum.AvatarSettingsLegacyCollisionMode,
	[83] = Enum.AvatarSettingsScaleMode,
	[84] = Enum.AvatarThumbnailCustomizationType,
	[85] = Enum.AvatarUnificationMode,
	[86] = Enum.Axis,
	[87] = Enum.BenefitType,
	[88] = Enum.BinType,
	[89] = Enum.BodyPart,
	[90] = Enum.BodyPartR15,
	[91] = Enum.BorderMode,
	[92] = Enum.BorderStrokePosition,
	[93] = Enum.BreakReason,
	[94] = Enum.BulkMoveMode,
	[95] = Enum.BundleType,
	[96] = Enum.Button,
	[97] = Enum.ButtonStyle,
	[98] = Enum.CageType,
	[99] = Enum.CameraMode,
	[100] = Enum.CameraPanMode,
	[101] = Enum.CameraType,
	[103] = Enum.CaptureGalleryPermission,
	[104] = Enum.CaptureType,
	[105] = Enum.CatalogCategoryFilter,
	[106] = Enum.CatalogSortAggregation,
	[107] = Enum.CatalogSortType,
	[108] = Enum.CellBlock,
	[109] = Enum.CellMaterial,
	[110] = Enum.CellOrientation,
	[111] = Enum.CenterDialogType,
	[112] = Enum.CharacterControlMode,
	[113] = Enum.ChatCallbackType,
	[114] = Enum.ChatColor,
	[115] = Enum.ChatMode,
	[116] = Enum.ChatPrivacyMode,
	[117] = Enum.ChatRestrictionStatus,
	[118] = Enum.ChatStyle,
	[119] = Enum.ChatVersion,
	[120] = Enum.ClientAnimatorThrottlingMode,
	[121] = Enum.CloseReason,
	[123] = Enum.CollisionFidelity,
	[124] = Enum.CommandPermission,
	[125] = Enum.CompositeValueCurveType,
	[126] = Enum.CompressionAlgorithm,
	[127] = Enum.ComputerCameraMovementMode,
	[128] = Enum.ComputerMovementMode,
	[129] = Enum.ConfigSnapshotErrorState,
	[130] = Enum.ConnectionError,
	[131] = Enum.ConnectionState,
	[132] = Enum.ContentSourceType,
	[133] = Enum.ContextActionPriority,
	[134] = Enum.ContextActionResult,
	[135] = Enum.ControlMode,
	[136] = Enum.CoreGuiType,
	[137] = Enum.CreateAssetResult,
	[138] = Enum.CreateContentResult,
	[139] = Enum.CreateOutfitFailure,
	[140] = Enum.CreatorType,
	[141] = Enum.CreatorTypeFilter,
	[142] = Enum.CurrencyType,
	[143] = Enum.CustomCameraMode,
	[144] = Enum.DataModelExtractorFileType,
	[145] = Enum.DataStoreRequestType,
	[146] = Enum.DebugBreakModeType,
	[147] = Enum.DebuggerResumeType,
	[148] = Enum.DevCameraOcclusionMode,
	[149] = Enum.DevComputerCameraMovementMode,
	[150] = Enum.DevComputerMovementMode,
	[151] = Enum.DevTouchCameraMovementMode,
	[152] = Enum.DevTouchMovementMode,
	[153] = Enum.DeveloperMemoryTag,
	[154] = Enum.DeviceFeatureType,
	[155] = Enum.DeviceForm,
	[156] = Enum.DeviceLevel,
	[157] = Enum.DeviceType,
	[158] = Enum.DialogBehaviorType,
	[159] = Enum.DialogPurpose,
	[160] = Enum.DialogTone,
	[161] = Enum.DigitsRigDescriptionSide,
	[162] = Enum.DiscountType,
	[163] = Enum.DisplayScalingMode,
	[164] = Enum.DisplaySize,
	[165] = Enum.DomainType,
	[166] = Enum.DominantAxis,
	[167] = Enum.DragDetectorDragStyle,
	[168] = Enum.DragDetectorPermissionPolicy,
	[169] = Enum.DragDetectorResponseStyle,
	[170] = Enum.DraggingScrollBar,
	[171] = Enum.EasingDirection,
	[172] = Enum.EasingStyle,
	[173] = Enum.EditableStatus,
	[174] = Enum.ElasticBehavior,
	[175] = Enum.EmitterPositionType,
	[176] = Enum.EngagementLevel,
	[177] = Enum.EngineFolder,
	[178] = Enum.EnviromentalPhysicsThrottle,
	[179] = Enum.ExperienceActivationStatus,
	[180] = Enum.ExperienceAuthScope,
	[181] = Enum.ExperienceEventStatus,
	[182] = Enum.ExperienceStateCaptureSelectionMode,
	[183] = Enum.ExperienceStateRecordingLoadMode,
	[184] = Enum.ExperienceStateRecordingLoadSourceType,
	[185] = Enum.ExperienceStateRecordingPlaybackMode,
	[186] = Enum.ExplosionType,
	[187] = Enum.FACSDataLod,
	[188] = Enum.FacialAgeEstimationResultType,
	[189] = Enum.FacialAnimationStreamingState,
	[190] = Enum.FacsActionUnit,
	[191] = Enum.FeatureRestrictionAbuseVector,
	[192] = Enum.FeedbackType,
	[193] = Enum.FieldOfViewMode,
	[194] = Enum.FillDirection,
	[195] = Enum.FilterResult,
	[196] = Enum.FilterType,
	[197] = Enum.FinishRecordingOperation,
	[198] = Enum.FluidFidelity,
	[199] = Enum.FluidForces,
	[200] = Enum.Font,
	[201] = Enum.FontSize,
	[202] = Enum.FontStyle,
	[203] = Enum.FontWeight,
	[204] = Enum.ForceLimitMode,
	[205] = Enum.FormFactor,
	[206] = Enum.FrameStyle,
	[207] = Enum.FriendRequestEvent,
	[208] = Enum.FriendStatus,
	[209] = Enum.FunctionalTestResult,
	[210] = Enum.GameAvatarType,
	[211] = Enum.GamepadType,
	[212] = Enum.GearGenreSetting,
	[213] = Enum.GearType,
	[214] = Enum.Genre,
	[215] = Enum.GradientTileMode,
	[216] = Enum.GradientType,
	[217] = Enum.GraphicsOptimizationMode,
	[218] = Enum.GroupMembershipStatus,
	[219] = Enum.GuiState,
	[220] = Enum.GuiType,
	[221] = Enum.HandlesStyle,
	[222] = Enum.HapticEffectType,
	[223] = Enum.HashAlgorithm,
	[224] = Enum.HighlightDepthMode,
	[225] = Enum.HorizontalAlignment,
	[226] = Enum.HttpCachePolicy,
	[227] = Enum.HttpCompression,
	[228] = Enum.HttpContentType,
	[229] = Enum.HttpError,
	[230] = Enum.HttpRequestType,
	[231] = Enum.HumanoidCollisionType,
	[232] = Enum.HumanoidDisplayDistanceType,
	[233] = Enum.HumanoidHealthDisplayType,
	[234] = Enum.HumanoidRigType,
	[235] = Enum.HumanoidStateType,
	[236] = Enum.IKCollisionsMode,
	[237] = Enum.IKControlConstraintSupport,
	[238] = Enum.IKControlType,
	[239] = Enum.IXPLoadingStatus,
	[240] = Enum.ImageAlphaType,
	[241] = Enum.ImageCombineType,
	[242] = Enum.InOut,
	[243] = Enum.InfoType,
	[244] = Enum.InitialDockState,
	[245] = Enum.InputActionType,
	[246] = Enum.InputBindingType,
	[247] = Enum.InputSink,
	[248] = Enum.InputType,
	[249] = Enum.InstanceFileSyncStatus,
	[250] = Enum.IntermediateMeshGenerationResult,
	[251] = Enum.InternalVideoUsage,
	[252] = Enum.InterpolationThrottlingMode,
	[253] = Enum.InviteState,
	[254] = Enum.ItemLineAlignment,
	[255] = Enum.JoinSource,
	[256] = Enum.JointCreationMode,
	[257] = Enum.KeyCode,
	[258] = Enum.KeyCodeStringFormat,
	[259] = Enum.KeyInterpolationMode,
	[260] = Enum.KeywordFilterType,
	[261] = Enum.Language,
	[262] = Enum.LeftRight,
	[263] = Enum.LightingStyle,
	[264] = Enum.Limb,
	[265] = Enum.LineJoinMode,
	[266] = Enum.ListenerLocation,
	[267] = Enum.ListenerPositionType,
	[268] = Enum.ListenerType,
	[271] = Enum.LoadCharacterLayeredClothing,
	[272] = Enum.LoadDynamicHeads,
	[273] = Enum.LocationType,
	[274] = Enum.LuauTypeCheckMode,
	[275] = Enum.MakeupType,
	[276] = Enum.MarketplaceBulkPurchasePromptStatus,
	[277] = Enum.MarketplaceItemPurchaseStatus,
	[278] = Enum.MarketplaceProductType,
	[279] = Enum.MatchmakingType,
	[280] = Enum.Material,
	[281] = Enum.MaterialPattern,
	[282] = Enum.MembershipType,
	[283] = Enum.MeshPartHeadsAndAccessories,
	[285] = Enum.MeshType,
	[286] = Enum.MessageType,
	[287] = Enum.ModelLevelOfDetail,
	[288] = Enum.ModelStreamingBehavior,
	[289] = Enum.ModelStreamingMode,
	[290] = Enum.ModerationResultCategory,
	[291] = Enum.ModerationResultLabel,
	[292] = Enum.ModerationStatus,
	[293] = Enum.ModifierKey,
	[294] = Enum.MouseBehavior,
	[295] = Enum.MoveState,
	[296] = Enum.MuteState,
	[297] = Enum.NameOcclusion,
	[298] = Enum.NegateOperationHiddenHistory,
	[299] = Enum.NetworkOwnership,
	[300] = Enum.NetworkStatus,
	[301] = Enum.NoiseType,
	[302] = Enum.NormalId,
	[303] = Enum.NotificationButtonType,
	[304] = Enum.OperationType,
	[305] = Enum.OrientationAlignmentMode,
	[306] = Enum.OutfitSource,
	[307] = Enum.OutfitType,
	[308] = Enum.OverrideMouseIconBehavior,
	[309] = Enum.PackagePermission,
	[310] = Enum.PartType,
	[311] = Enum.ParticleEmitterShape,
	[312] = Enum.ParticleEmitterShapeInOut,
	[313] = Enum.ParticleEmitterShapeStyle,
	[314] = Enum.ParticleFlipbookLayout,
	[315] = Enum.ParticleFlipbookMode,
	[316] = Enum.ParticleFlipbookTextureCompatible,
	[317] = Enum.ParticleOrientation,
	[318] = Enum.PathStatus,
	[319] = Enum.PathWaypointAction,
	[320] = Enum.PathfindingUseImprovedSearch,
	[321] = Enum.PeoplePageLayout,
	[323] = Enum.PhysicsSimulationRate,
	[324] = Enum.PhysicsSteppingMethod,
	[327] = Enum.Platform,
	[328] = Enum.PlaybackState,
	[329] = Enum.PlayerActions,
	[330] = Enum.PlayerCharacterDestroyBehavior,
	[331] = Enum.PlayerChatType,
	[332] = Enum.PlayerDataErrorState,
	[333] = Enum.PlayerDataLoadFailureBehavior,
	[334] = Enum.PlayerExitReason,
	[335] = Enum.PlayerPlatformActivationStatus,
	[336] = Enum.PlayerPlatformSpenderStatus,
	[337] = Enum.PoseEasingDirection,
	[338] = Enum.PoseEasingStyle,
	[339] = Enum.PositionAlignmentMode,
	[340] = Enum.PredictionMode,
	[341] = Enum.PredictionStatus,
	[342] = Enum.PreferredInput,
	[343] = Enum.PreferredTextSize,
	[344] = Enum.PrimalPhysicsSolver,
	[345] = Enum.PrimitiveType,
	[346] = Enum.PrivilegeType,
	[347] = Enum.ProductLocationRestriction,
	[348] = Enum.ProductPurchaseChannel,
	[349] = Enum.ProductPurchaseDecision,
	[350] = Enum.PromptCreateAssetResult,
	[351] = Enum.PromptCreateAvatarResult,
	[352] = Enum.PromptExperienceDetailsResult,
	[353] = Enum.PromptLinkSharingResult,
	[354] = Enum.PromptPublishAssetResult,
	[355] = Enum.PropertyStatus,
	[356] = Enum.ProximityPromptExclusivity,
	[357] = Enum.ProximityPromptInputType,
	[358] = Enum.ProximityPromptStyle,
	[359] = Enum.PurchaseOption,
	[360] = Enum.R15CollisionType,
	[361] = Enum.RaycastFilterType,
	[362] = Enum.ReadCapturesFromGalleryResult,
	[363] = Enum.ReceiptDecision,
	[364] = Enum.ReceiptType,
	[365] = Enum.RecommendationActionType,
	[366] = Enum.RecommendationDepartureIntent,
	[367] = Enum.RecommendationImpressionType,
	[368] = Enum.RecommendationItemContentType,
	[369] = Enum.RecommendationItemVisibility,
	[370] = Enum.RecommendationPreferenceTargetType,
	[371] = Enum.RecommendationPreferenceType,
	[372] = Enum.RejectCharacterDeletions,
	[373] = Enum.RenderFidelity,
	[374] = Enum.RenderPriority,
	[375] = Enum.RenderingCacheOptimizationMode,
	[376] = Enum.RenderingTestComparisonMethod,
	[377] = Enum.ReplicateInstanceDestroySetting,
	[378] = Enum.ResamplerMode,
	[379] = Enum.ReservedHighlightId,
	[381] = Enum.ReturnKeyType,
	[382] = Enum.ReverbType,
	[383] = Enum.ReviewableContentState,
	[384] = Enum.RibbonTool,
	[385] = Enum.RigLabel,
	[388] = Enum.RollOffMode,
	[389] = Enum.RolloutState,
	[390] = Enum.RotationOrder,
	[391] = Enum.RotationType,
	[392] = Enum.RsvpStatus,
	[393] = Enum.RtlTextSupport,
	[394] = Enum.RunContext,
	[395] = Enum.RunState,
	[396] = Enum.RuntimeUndoBehavior,
	[397] = Enum.SafeAreaCompatibility,
	[398] = Enum.SalesTypeFilter,
	[399] = Enum.SandboxedInstanceMode,
	[400] = Enum.SaveAvatarThumbnailCustomizationFailure,
	[401] = Enum.SaveFilter,
	[402] = Enum.SavedQualitySetting,
	[403] = Enum.ScaleType,
	[404] = Enum.ScopeCheckResult,
	[405] = Enum.ScreenInsets,
	[406] = Enum.ScreenOrientation,
	[407] = Enum.ScreenshotCaptureResult,
	[408] = Enum.ScriptStoppedReason,
	[409] = Enum.ScriptVariableScope,
	[410] = Enum.ScrollBarInset,
	[411] = Enum.ScrollingDirection,
	[412] = Enum.SecurityCapability,
	[413] = Enum.SelectionBehavior,
	[414] = Enum.SelectionRenderMode,
	[415] = Enum.SelfViewPosition,
	[416] = Enum.SensorMode,
	[417] = Enum.SensorUpdateType,
	[419] = Enum.ShowAdResult,
	[420] = Enum.SignalBehavior,
	[421] = Enum.SimulationMode,
	[422] = Enum.SizeConstraint,
	[423] = Enum.SlimTintMode,
	[424] = Enum.SolidPrimitiveType,
	[425] = Enum.SolverConvergenceMetricType,
	[426] = Enum.SolverConvergenceVisualizationMode,
	[427] = Enum.SortDirection,
	[428] = Enum.SortOrder,
	[429] = Enum.SpecialKey,
	[430] = Enum.StartCorner,
	[431] = Enum.StateObjectFieldType,
	[432] = Enum.Status,
	[433] = Enum.StepFrequency,
	[434] = Enum.StreamOutBehavior,
	[435] = Enum.StreamingIntegrityMode,
	[436] = Enum.StreamingPauseMode,
	[437] = Enum.StrokeSizingMode,
	[438] = Enum.StudioAction,
	[439] = Enum.StudioCaptureScreenshotFormat,
	[440] = Enum.StudioCloseMode,
	[441] = Enum.StudioDataModelType,
	[442] = Enum.StudioPlaceUpdateFailureReason,
	[443] = Enum.Style,
	[444] = Enum.SubscriptionExpirationReason,
	[445] = Enum.SubscriptionPaymentStatus,
	[446] = Enum.SubscriptionPeriod,
	[447] = Enum.SubscriptionState,
	[448] = Enum.SurfaceConstraint,
	[449] = Enum.SurfaceGuiShape,
	[450] = Enum.SurfaceGuiSizingMode,
	[451] = Enum.SurfaceType,
	[452] = Enum.SwipeDirection,
	[453] = Enum.SystemThemeValue,
	[454] = Enum.TableMajorAxis,
	[457] = Enum.Technology,
	[458] = Enum.TelemetryBackend,
	[459] = Enum.TelemetryStandardizedField,
	[460] = Enum.TeleportMethod,
	[461] = Enum.TeleportResult,
	[462] = Enum.TeleportState,
	[463] = Enum.TeleportType,
	[464] = Enum.TerrainAcquisitionMethod,
	[465] = Enum.TerrainFace,
	[466] = Enum.TerrainLiquidMergeOperation,
	[467] = Enum.TerrainSolidMergeOperation,
	[468] = Enum.TextChatMessageStatus,
	[469] = Enum.TextDirection,
	[470] = Enum.TextFilterContext,
	[471] = Enum.TextInputType,
	[472] = Enum.TextTruncate,
	[473] = Enum.TextXAlignment,
	[474] = Enum.TextYAlignment,
	[475] = Enum.TextureMode,
	[476] = Enum.TextureQueryType,
	[477] = Enum.ThreadPoolConfig,
	[478] = Enum.ThrottlingPriority,
	[479] = Enum.ThumbnailSize,
	[480] = Enum.ThumbnailType,
	[481] = Enum.TickCountSampleMethod,
	[482] = Enum.TonemapperPreset,
	[483] = Enum.TopBottom,
	[484] = Enum.TouchCameraMovementMode,
	[485] = Enum.TouchMovementMode,
	[486] = Enum.TrackerError,
	[487] = Enum.TrackerExtrapolationFlagMode,
	[488] = Enum.TrackerFaceTrackingStatus,
	[489] = Enum.TrackerLodFlagMode,
	[490] = Enum.TrackerLodValueMode,
	[491] = Enum.TrackerMode,
	[492] = Enum.TrackerPromptEvent,
	[493] = Enum.TrackerType,
	[494] = Enum.TriStateBoolean,
	[495] = Enum.TweenStatus,
	[496] = Enum.UICaptureMode,
	[497] = Enum.UIDragDetectorBoundingBehavior,
	[498] = Enum.UIDragDetectorDragRelativity,
	[499] = Enum.UIDragDetectorDragSpace,
	[500] = Enum.UIDragDetectorDragStyle,
	[501] = Enum.UIDragDetectorResponseStyle,
	[502] = Enum.UIDragSpeedAxisMapping,
	[503] = Enum.UIFlexAlignment,
	[504] = Enum.UIFlexMode,
	[505] = Enum.UiMessageType,
	[506] = Enum.UpdateState,
	[507] = Enum.UploadCaptureResult,
	[508] = Enum.UsageContext,
	[509] = Enum.UserCFrame,
	[510] = Enum.UserInputState,
	[511] = Enum.UserInputType,
	[512] = Enum.VRComfortSetting,
	[513] = Enum.VRControllerModelMode,
	[514] = Enum.VRDeviceType,
	[515] = Enum.VRLaserPointerMode,
	[516] = Enum.VRSafetyBubbleMode,
	[517] = Enum.VRScaling,
	[518] = Enum.VRSessionState,
	[519] = Enum.VRTouchpad,
	[520] = Enum.VRTouchpadMode,
	[521] = Enum.VelocityConstraintMode,
	[522] = Enum.VerticalAlignment,
	[523] = Enum.VerticalScrollBarPosition,
	[524] = Enum.VibrationMotor,
	[525] = Enum.VideoCaptureResult,
	[526] = Enum.VideoCaptureStartedResult,
	[527] = Enum.VideoDeviceCaptureQuality,
	[528] = Enum.VideoError,
	[529] = Enum.VideoSampleSize,
	[530] = Enum.VirtualCursorMode,
	[531] = Enum.VoiceChatDistanceAttenuationType,
	[532] = Enum.VoiceChatState,
	[533] = Enum.VoiceClientLeaveReasons,
	[534] = Enum.VoiceControlPath,
	[535] = Enum.VoiceRccReconnectReason,
	[536] = Enum.VolumetricAudio,
	[537] = Enum.WaterDirection,
	[538] = Enum.WaterForce,
	[539] = Enum.WebSocketState,
	[540] = Enum.WebStreamClientState,
	[541] = Enum.WebStreamClientType,
	[542] = Enum.WeldConstraintPreserve,
	[543] = Enum.WhenUserFirstPlayed,
	[544] = Enum.WhisperChatPrivacyMode,
	[545] = Enum.WrapLayerAutoSkin,
	[546] = Enum.WrapLayerDebugMode,
	[547] = Enum.WrapTargetDebugMode,
	[548] = Enum.ZIndexBehavior,
}

local DecoderModule = {
	Type = 0x07,
}

local function findByValue(enum: Enum, value: number): EnumItem?
	for _, item in enum:GetEnumItems() do
		if item.Value == value then
			return item
		end
	end
	return nil
end

local function findUniqueByValue(value: number): EnumItem?
	local match: EnumItem? = nil
	for _, enum in Enum:GetEnums() do
		for _, item in enum:GetEnumItems() do
			if item.Value ~= value then
				continue
			end

			if match then
				return nil -- ambiguous across enum families
			end

			match = item
		end
	end

	return match
end

function DecoderModule.Parse(data: buffer, offset: number, _t: number, _parser: any, _depth: number)
	local typeId = buffer.readu16(data, offset)
	local value, newOffset = Codec.LEB128(data, offset + 2)

	local enum = TYPE_ID_TO_ENUM[typeId]
	local item = if enum then findByValue(enum, value) else nil
	if not item then
		item = findUniqueByValue(value)
		if item and not TYPE_ID_TO_ENUM[typeId] then
			TYPE_ID_TO_ENUM[typeId] = item.EnumType
		end
	end

	return {
		Result = item or value,
		Offset = newOffset,
	}
end

return DecoderModule
